core/src/main/scala/akka/persistence/cassandra/query/TagViewSequenceNumberScanner.scala

```diff
@@ -7,80 +7,111 @@ package akka.persistence.cassandra.query
 
-      def doIt(): Future[Map[Tag, (TagPidSequenceNr, UUID)]] = {
-        val bound = ps.bind(tag, bucket.key: JLong, offset, to).setExecutionProfileName(profile)
+    def doIt(): Future[Map[PersistenceId, (TagPidSequenceNr, UUID)]] = {
+
+      // How many buckets is this spread across?
+      val startBucket = TimeBucket(fromOffset, bucketSize)
+      val endBucket = TimeBucket(toOffset, bucketSize)
+
+      require(startBucket <= endBucket)
+
+      if (log.isDebugEnabled) {
         log.debug(
-          "Scanning tag: {} bucket: {}, from: {}, to: {}",
-          tag,
-          bucket.key,
-          formatOffset(offset),
-          formatOffset(to))
-        val source = session.select(bound)
-        val doneIt = source
-          .map(row => (row.getString("persistence_id"), row.getLong("tag_pid_sequence_nr"), row.getUuid("timestamp")))
-          .runFold(Map.empty[Tag, (TagPidSequenceNr, UUID)]) {
-            case (acc, (pid, tagPidSequenceNr, timestamp)) =>
-              val (newTagPidSequenceNr, newTimestamp) = acc.get(pid) match {
-                case None =>
-                  (tagPidSequenceNr, timestamp)
-                case Some((currentTagPidSequenceNr, currentTimestamp)) =>
-                  if (tagPidSequenceNr < currentTagPidSequenceNr)
-                    (tagPidSequenceNr, timestamp)
-                  else
-                    (currentTagPidSequenceNr, currentTimestamp)
-              }
-              acc + (pid -> ((newTagPidSequenceNr, newTimestamp)))
+          s"Scanning tag: $tag from: {}, to: {}. Bucket {} to {}",
+          formatOffset(fromOffset),
+          formatOffset(toOffset),
+          startBucket,
+          endBucket)
+      }
+
+      Source
+        .unfold(startBucket)(current => {
+          if (current <= endBucket) {
+            Some((current.next(), current))
+          } else {
+            None
           }
-        doneIt.flatMap { result =>
+        })
+        .flatMapConcat(bucket => {
+          log.debug("Scanning bucket {}", bucket)
+          session.selectTagSequenceNrs(tag, bucket, fromOffset, toOffset)
+        })
+        .map(row => (row.getString("persistence_id"), row.getLong("tag_pid_sequence_nr"), row.getUuid("timestamp")))
+        .runFold(Map.empty[Tag, (TagPidSequenceNr, UUID)]) {
+          case (acc, (pid, tagPidSequenceNr, timestamp)) =>
+            val (newTagPidSequenceNr, newTimestamp) = acc.get(pid) match {
+              case None =>
+                (tagPidSequenceNr, timestamp)
+              case Some((currentTagPidSequenceNr, currentTimestamp)) =>
+                if (whichToKeep(tagPidSequenceNr, currentTagPidSequenceNr) == tagPidSequenceNr)
+                  (tagPidSequenceNr, timestamp)
+                else
+                  (currentTagPidSequenceNr, currentTimestamp)
+            }
+            acc + (pid -> ((newTagPidSequenceNr, newTimestamp)))
+        }
+        .flatMap { result =>
           if (deadline.hasTimeLeft()) {
             doIt()
```

Comment:
I see that the `scanningPeriod` isn't used (Duration.Zero) for the backtracking scanning, but I guess it's used for new-persistence-id-scan-timeout and offset-scanning-period. What is the purpose of purpose of performing the queries and then throw away the result here compared to just waiting for that duration and then perform one scanning query?

Reply:
in case the scan takes a longer than the `offset-scanning-period`, it is set to 200ms by default and for real datasets the query will take that long. If we only started it after the 200ms then the total time would be 200ms + query time, rather than query time.

Reply:
Created issue https://github.com/akka/akka-persistence-cassandra/issues/621 for future consideration.